### ISSUE TITLE
Allow configuring the APP_BASE via host_fxr_main

### ISSUE THREAD

**Issue by davidfowl** [dotnet/core-setup#3896]
We're not hosting .NET Core in the IIS worker process in 2.1 and as a result we'd like to make the application base match the application path. See https://github.com/aspnet/IISIntegration/issues/725.

Is there any way to tweak the CLR settings today via an API call?

/cc @jkotalik @pakrym

**Comment by jkotas** [comment]
How does ` \\?\` get into `AppDomain.CurrentDomain.BaseDirectory` under IIS ?

@JeremyKuhne Do you think that we should be always stripping ` \\?\` from `AppDomain.CurrentDomain.BaseDirectory` ? Or is it expected that the callers of various framework APIs that return paths are prepared to deal with ` \\?\` on Windows?

**Comment by davidfowl** [comment]
That part I'm not sure about but even without that syntax the app base is incorrect. We'd like to be able to set it properly.

**Comment by jkotas** [comment]
You can set the AppDomain base directory to anything you want using `AppContext.SetData("APP_CONTEXT_BASE_DIRECTORY", ...)`. 

This sounds like the general problem that you would like to run your managed code before the apps Main method is invoked. Do you have a way to do that in the ASP.NET IIS host today?

**Comment by JeremyKuhne** [comment]
> Do you think that we should be always stripping \\?\

I think you should probably strip it, particularly if the path doesn't end in a period or space. Outside of that everything else should work fine.

**Comment by davidfowl** [comment]
> You can set the AppDomain base directory to anything you want using AppContext.SetData("APP_CONTEXT_BASE_DIRECTORY", ...).

No I can't it's too late. I actually don't want to run any managed code. I just want to configure the host before my main runs.

> This sounds like the general problem that you would like to run your managed code before the apps Main method is invoked. Do you have a way to do that in the ASP.NET IIS host today?

I don't want to revert to this yet. The CoreCLR hosting APIs lets us set the app base, we just need a way to set it from host_fxr_main.

**Comment by jkotas** [comment]
The argument for `hostfxr_main` is the command line. Are you asking for a new command line option for `dotnet exec` that lets you set appbase?

**Comment by davidfowl** [comment]
Yes or a new API that lets us set args + host settings.

**Comment by pakrym** [comment]
The command line argument would be less desirable because then we would have to check that user didn't set it on his own and get into the business of argument parsing.

**Comment by jkotalik** [comment]
At the same time though, I don't want to add unnecessary bloat to hostfxr, like a simple call to set the BaseDirectory on the AppDirectory.

**Comment by pakrym** [comment]
We can just add a native alternative to `AppContext.SetData`

**Comment by jkotas** [comment]
What does the user specified command line option for IIS hosting looks like today? Can users provide the host command line options?

Note that this would be host option (the one that comes before the .dll name to execute). dotnet exec has host command line options to configure close to everything imaginable: 

```
host-options:
  --additionalprobingpath <path>      Path containing probing policy and assemblies to probe for
  --depsfile <path>                   Path to <application>.deps.json file
  --runtimeconfig <path>              Path to <application>.runtimeconfig.json file
  --fx-version <version>              Version of the installed Shared Framework to use to run the application.
  --roll-forward-on-no-candidate-fx   Roll forward on no candidate shared framework is enabled
  --additional-deps <path>            Path to additonal deps.json file
```

**Comment by pakrym** [comment]
> Can users provide the host command line options?

Yes, we are passing contents of arguments property from web.config directly to `hostfxr_main` so it can contain host arguments.

**Comment by pakrym** [comment]
The `\\?\` part is coming from ANCM calculating absolute path. 

@davidfowl wants to change it to point to IIS site location instead of entry dll location.

**Comment by jkotalik** [comment]
When we launch w3wp.exe, it has the current working directory to C:\Windows\System32\inetsrv, which is what is returned by Environment.CurrentDirectory. 
For the AppDomain itself, it does return the correct path to the application. We can strip the \\\\?\\ portion.

**Comment by jkotas** [comment]
I do not think you should be in the business of stripping `\\?\` portion from appbase. The same problem exists for simple command line activation. Try: `dotnet \\?\C:\HelloWorld\bin\Debug\netcoreapp2.0\publish\HelloWorld.dll`. This should be fixed in a more central place so that the fix applies to everybody, not just for IIS hosting.

Do you still think you need the API to set the appbase, assuming we will figure out the central place where to handle `\\?\` ?

**Comment by steveharter** [comment]
Other properties values are passed in via the runtimeconfig:
```
    "configProperties": {
      "System.GC.Server": true
    }
```

If we could use this mechanism, then the hostfxr API entry point addition wouldn't need to change, and the code to pull in this setting would just be in hostpolicy.

However I assume we don't want to leverage this mechanism because you don't want to write the appbase setting to the runtimeconfig, correct?

**Comment by jkotas** [comment]
I have fixed the title to what I believe the actual problem is. It is not specific to IIS. To reproduce the problem

1. Create console app at c:\HelloWorld
2. Change the code to (repro from https://github.com/aspnet/IISIntegration/issues/725):
```
var fileName = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "test.xml");
var doc = new XmlDocument();
doc.Load(fileName);
```
3. Publish it and run it using `dotnet \\?\C:\HelloWorld\bin\Debug\netcoreapp2.0\publish\HelloWorld.dll`

Result: System.UriFormatException: Invalid URI: The hostname could not be parsed.

**Comment by steveharter** [comment]
So implementation options as I see them:
1) Add a new entry point to hostfxr (e.g. `hostfxr_main_with_clr_properties`) that takes the information, probably a name-value pair (e.g. "APP_CONTEXT_BASE_DIRECTORY", "C:\\foo")
2) Add a new host option (e.g. `--appbase`)
3) Write a setting (e.g. `appBase`) to the runtimeconfig in the `configProperties`

If option 3 is acceptable to ASP.NET then that is the least intrusive to the host (only requires hostpolicy changes, and not "contract" additions between hostfxr and hostpolicy); easiest to implement.

Option 2 would expose it to everyone, which probably isn't a good idea.

Option 1 something like:
```
SHARED_API int32_t hostfxr_main_with_clr_properties(const int argc, const pal::char_t* argv[], const int property_count, const pal::char_t* property_names[], const pal::char_t* property_values[])
```

**Comment by jkotas** [comment]
@steveharter The actual problem from the linked bug is that code out there chokes on paths starting with `\\?` . I do not think that this problem should be fixed by providing an API to set appbase path. 

I think that the best solutions to this problem is to strip `\\?` from the paths when it does not need to be there. This does not require any host API changes. We just need to find the right place to do it.

**Comment by steveharter** [comment]
OK, so I assume the original "make the application base match the application path" requirement no longer holds.

From the preview code [link ](https://github.com/aspnet/IISIntegration/blob/b70191fb82bef2346a931a3c645f5cb8132c8a79/src/IISLib/util.cxx#L7) it appears `\\?\UNC\` is another variant to account for.

**Comment by steveharter** [comment]
I'm assuming we need this for 2.1.0. Correct?

**Comment by pakrym** [comment]
Yes

**Comment by steveharter** [comment]
Where will the `\\?` and `\\?\UNC\` be stripped at?
- In the host
- In the CLR
- By ASP.NET

**Comment by karelz** [comment]
Noe enough info for it to be actionable yet, marking as Post-ZBB.

**Comment by pakrym** [comment]
ASP.NET is out of equation here (repro doesn't even have it https://github.com/dotnet/core-setup/issues/3896#issuecomment-377055455). So it has to be either host of clr

**Comment by karelz** [comment]
@pakrym why is it important for 2.1? Is it a regression from 2.0? Were there new features added that depend on it?

**Comment by pakrym** [comment]
@karelz it's not a regression but paths starting with `\\?\` are completely valid windows paths that are handled incorrectly by the host.

We hit it in AspNetCore IIS module scenario where it was normalizing all paths by `\\?\` before calling dotnet and caused all kinds of failures in different parts of the framework (Uri handling for example)

**Comment by karelz** [comment]
Is the scenario more common / more important in 2.1 than in 2.0?
Given that it came late and we do not have agreement on solution, I'd like to understand why it is important to fix it in 2.1. How should I compare it to other 2.1-critical work?

**Comment by karelz** [comment]
Moving to Future - it does not seem to be critical for 2.1.

**Comment by davidfowl** [comment]
Yes we can punt this for now.
